### PR TITLE
Implement CandidateController methods

### PR DESCRIPTION
**Pull request by andriimanzar**

### CODE REVIEW

**Comment by knurff** [comment]
|File|Coverage [31%]|:green_apple:|
|:-|:-:|:-:|
|[TimeConverter.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FTimeConverter.java)|100%|:green_apple:|
|[InvalidCandidateSlotDateException.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FInvalidCandidateSlotDateException.java)|100%|:green_apple:|
|[CandidateSlotValidator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2Fvalidator%2FCandidateSlotValidator.java)|95.65%|:green_apple:|
|[CandidateService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FCandidateService.java)|77.38%|:green_apple:|
|[CandidateSlot.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FCandidateSlot.java)|46.58%|:green_apple:|
|[TimePeriod.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FTimePeriod.java)|12.1%|:green_apple:|
|[CandidateSlotMapper.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FCandidateSlotMapper.java)|0%|:x:|
|[CandidateSlotDto.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2Fdto%2FCandidateSlotDto.java)|0%|:x:|
|[ApplicationExceptionHandler.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FApplicationExceptionHandler.java)|0%|:x:|
|[SlotContainsBookingsException.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FSlotContainsBookingsException.java)|0%|:x:|
|[WebSecurityConfig.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fsecurity%2FWebSecurityConfig.java)|0%|:x:|
|[CandidateController.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/9dd753494e39b20a71ccf059aee0cf094bc286a5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2FCandidateController.java)|0%|:x:|

|Total Project Coverage|20.1%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/intellias/intellistart/interviewplanning/controller/CandidateController.java [hunk 1]
```diff
@@ -1,11 +1,61 @@
...
 /**
  * CandidateController controller.
  */
 @RestController
+@RequestMapping("/candidates/current/slots")
 public class CandidateController {
 
+  private final CandidateService candidateService;
+
+  @Autowired
```
[the remark this answers is not in the document]
Comment (Famezzs) [reply src/main/java/com/intellias/intellistart/interviewplanning/controller/CandidateController.java:32]: Просто, якщо я правильно пам'ятаю, можна й без нього, хоча можу помилятися.

[the remark this answers is not in the document]
Comment (andriimanzar) [reply src/main/java/com/intellias/intellistart/interviewplanning/controller/CandidateController.java:32]: Можна, але краще явно вказувати

File: src/main/java/com/intellias/intellistart/interviewplanning/controller/CandidateController.java [hunk 2]
```diff
@@ -1,11 +1,61 @@
...
+  }
+
+  @PutMapping("/{id}")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void editSlot(@PathVariable Long id, @RequestBody CandidateSlotDto candidateSlotDto) {
+    candidateSlotDto.setEmail(getUserEmailFromToken());
+    candidateService.editSlot(mapDtoToEntity(candidateSlotDto), id);
+  }
+
+  private String getUserEmailFromToken() {
```
[the remark this answers is not in the document]
Comment (andriimanzar) [reply src/main/java/com/intellias/intellistart/interviewplanning/controller/CandidateController.java:57]: Це тимчасовий метод, поки відключене Spring Security

File: src/main/java/com/intellias/intellistart/interviewplanning/service/validator/CandidateSlotValidator.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      throw new InvalidTimeSlotBoundariesException("Candidate slot time period must be in future");
+    }
+  }
+
+  private static boolean slotDateIsInThePastCondition(CandidateSlot candidateSlot) {
+    return candidateSlot.getDate().isBefore(LocalDate.now());
+  }
+
+  private static boolean slotTimeIsInThePastCondition(TimePeriod timePeriod) {
+    return timePeriod.getStartTime().isBefore(LocalTime.now()) || timePeriod.getEndTime()
```
[the remark this answers is not in the document]
Comment (andriimanzar) [reply src/main/java/com/intellias/intellistart/interviewplanning/service/validator/CandidateSlotValidator.java:55]: Перевіряється, чи стартовий АБО кінцевий час в минулому.  Навіть якщо один із параметрів в минулому - слот не валідний

**Comment by knurff** [comment]
|File|Coverage [31.13%]|:green_apple:|
|:-|:-:|:-:|
|[TimeConverter.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FTimeConverter.java)|100%|:green_apple:|
|[InvalidCandidateSlotDateException.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FInvalidCandidateSlotDateException.java)|100%|:green_apple:|
|[CandidateSlotValidator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2Fvalidator%2FCandidateSlotValidator.java)|95.71%|:green_apple:|
|[CandidateService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FCandidateService.java)|77.38%|:green_apple:|
|[CandidateSlot.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FCandidateSlot.java)|46.58%|:green_apple:|
|[TimePeriod.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FTimePeriod.java)|12.1%|:green_apple:|
|[RequestParser.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FRequestParser.java)|0%|:x:|
|[CandidateSlotMapper.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FCandidateSlotMapper.java)|0%|:x:|
|[CandidateSlotDto.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2Fdto%2FCandidateSlotDto.java)|0%|:x:|
|[ApplicationExceptionHandler.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FApplicationExceptionHandler.java)|0%|:x:|
|[SlotContainsBookingsException.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fexception%2FSlotContainsBookingsException.java)|0%|:x:|
|[WebSecurityConfig.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fsecurity%2FWebSecurityConfig.java)|0%|:x:|
|[CandidateController.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/1eb4ab6bcb1a5f556271278f43a5934a58bd2e6a/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2FCandidateController.java)|0%|:x:|

|Total Project Coverage|20.18%|:green_apple:|
|:-|:-:|:-:|

**Review by mariiasirenko2 — APPROVED** [review]
Approve

**Pull request merged**